### PR TITLE
[SPARK-12044] [SparkR] Fix usage of isnan, isNaN

### PR DESCRIPTION
**Pull request by yanboliang**
1, Add `isNaN` to `Column` for SparkR. `Column` should has three related variable functions: `isNaN, isNull, isNotNull`.
2, Replace `DataFrame.isNaN` with `DataFrame.isnan` at SparkR side. Because `DataFrame.isNaN` has been deprecated and will be removed at Spark 2.0.
<del>3, Add `isnull` to `DataFrame` for SparkR. `DataFrame` should has two related functions: `isnan, isnull`.<del>

cc @shivaram @sun-rui @felixcheung

### CODE REVIEW

File: R/pkg/inst/tests/test_sparkSQL.R [hunk 1]
```diff
@@ -878,7 +878,7 @@ test_that("column functions", {
   c2 <- avg(c) + base64(c) + bin(c) + bitwiseNOT(c) + cbrt(c) + ceil(c) + cos(c)
   c3 <- cosh(c) + count(c) + crc32(c) + exp(c)
   c4 <- explode(c) + expm1(c) + factorial(c) + first(c) + floor(c) + hex(c)
-  c5 <- hour(c) + initcap(c) + isNaN(c) + last(c) + last_day(c) + length(c)
+  c5 <- hour(c) + initcap(c) + isnan(c) + isnull(c) + last(c) + last_day(c) + length(c)
```
Comment (sun-rui) [inline R/pkg/inst/tests/test_sparkSQL.R]: please add test cases for "isNaN", "isNull", "isNotNul" for Column

**Comment by sun-rui** [comment]
LGTM except that it's good to add additional test cases.

File: R/pkg/R/functions.R [hunk 2]
```diff
@@ -488,19 +488,35 @@ setMethod("initcap",
...
-#' @rdname isNaN
-#' @name isNaN
+#' @rdname isnan
+#' @name isnan
 #' @family normal_funcs
 #' @export
-#' @examples \dontrun{isNaN(df$c)}
-setMethod("isNaN",
+#' @examples \dontrun{isnan(df$c)}
+setMethod("isnan",
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: should this be called (or have an alias) `is.nan`?
https://stat.ethz.ch/R-manual/R-devel/library/base/html/is.finite.html

File: R/pkg/R/functions.R [hunk 3]
```diff
@@ -488,19 +488,35 @@ setMethod("initcap",
...
+#' isnull
+#'
+#' Return true if the column is NULL.
+#'
+#' @rdname isnull
+#' @name isnull
+#' @family normal_funcs
+#' @export
+#' @examples \dontrun{isnull(df$c)}
+setMethod("isnull",
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: similarly, `is.null`?
https://stat.ethz.ch/R-manual/R-devel/library/base/html/NULL.html

File: R/pkg/R/generics.R [hunk 4]
```diff
@@ -623,6 +623,10 @@ setGeneric("getItem", function(x, ...) { standardGeneric("getItem") })
 
 #' @rdname column
 #' @export
+setGeneric("isNaN", function(x) { standardGeneric("isNaN") })
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: if `isNaN` is being removed, we shouldn't set this generic right?

Comment (felixcheung) [inline R/pkg/R/generics.R]: also, isNaN is around since Spark 1.5 - if we are taking this out we would need to note this breaking change in release doc with a JIRA

Comment (felixcheung) [inline R/pkg/R/generics.R]: sorry, rereading the description. it sounds like we have
`isNaN` for Column
`isnan` for DataFrame
?
That seems a bit confusing..

File: R/pkg/R/functions.R (hunk 2 above)
Comment (sun-rui) [inline R/pkg/R/functions.R]: yes, it makes sense to add an alias as is.nan

File: R/pkg/R/functions.R (hunk 3 above)
Comment (sun-rui) [inline R/pkg/R/functions.R]: is.null works on the object as a whole. is.na is better()?
```
> is.null(list(1, NULL))
[1] FALSE
> is.na(list(1, NA))
[1] FALSE  TRUE
```

Comment (yanboliang) [inline R/pkg/R/functions.R]: Indeed, ```is.null``` works on the whole object, but ```is.na``` has different semantics with ```is.null```. Should we still support the alias?

File: R/pkg/R/generics.R (hunk 4 above)
Comment (yanboliang) [inline R/pkg/R/generics.R]: Yes, indeed. I have send #10056 to push SQL side make change to uniform interface.

File: R/pkg/R/functions.R (hunk 3 above)
Comment (sun-rui) [inline R/pkg/R/functions.R]: In the context of DataFrame column, "null" means missing value, which I think NA in R means. When we read a column from a DataFrame to R side, null will be converted to NA, see the code at https://github.com/apache/spark/blob/master/R/pkg/R/deserialize.R#L115

Comment (felixcheung) [inline R/pkg/R/functions.R]: this gets tricky... R's usage for NULL is not the same as JVM's. In many cases it might be closer to R's NA though not exactly the same either.

eg. can't set a column in data.frame to NULL in the conventional way:
http://www.cookbook-r.com/Manipulating_data/Adding_and_removing_columns_from_a_data_frame/

Maybe it's fine to leave it as `isnull` for distinction - we really need some explanations on what is a NULL column.

Comment (felixcheung) [inline R/pkg/R/functions.R]: Our replies crossed - it would seem to me to say Scala/Python `isnull` == R `is.na` is confusing.

In any case we could definitely use some explanations on NULL/NA in the SPARKR programming guide, opened JIRA https://issues.apache.org/jira/browse/SPARK-12071

Comment (sun-rui) [inline R/pkg/R/functions.R]: good. It's a place where we can discuss how we map R NA or NULL to Scala null.

File: R/pkg/R/generics.R (hunk 4 above)
Comment (yanboliang) [inline R/pkg/R/generics.R]: It won't uniform the ```isNaN``` and ```isnan``` interface at SQL side, please refer the comments at #10056. Further more, there are different semantics between Scala and Java about ```isnull```. So let's narrow this PR to provide the same functions as Scala at SparkR side, this is my original motivation. I will add test cases for column as @sun-rui suggested. Then I think this PR can be merged firstly and we start to discuss corresponding alias or explanations about the ```NULL/NA``` difference. @sun-rui @felixcheung @shivaram

Comment (yanboliang) [inline R/pkg/R/generics.R]: And acturally, ```isnan``` and ```isNaN``` has do same thing at SparkR side only call different functions at Scala side.

**Comment by sun-rui** [comment]
@yanboliang, I am ok that we leave alias of is.na to  SPARK-12071. But it would be better you can add alias is.nan.

File: R/pkg/R/generics.R (hunk 4 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: This is loosely related to this PR and I'd suggest we discuss more on this. Currently this behavior is very confusing, see my example below @shivaram your thought?

tl;dr - it prints out value as NA but one would interact with it as NULL
```
> head(a)
    area     peri     shape   perm
1   4990 2791.900 0.0903296    6.3
2   7002 3892.600 0.1486220    6.3
3   7558       NA 0.1833120    6.3
4   7352 3869.320 0.1170630    6.3
> df <- as.DataFrame(sqlContext, a)
> head(df)
  area    peri     shape perm
1 4990 2791.90 0.0903296  6.3
2 7002 3892.60 0.1486220  6.3
3 7558      NA 0.1833120  6.3
4 7352 3869.32 0.1170630  6.3
5 7943 3948.54 0.1224170 17.1
6 7979 4010.15 0.1670450 17.1
> a <- filter(df, "peri != NA")
15/12/02 19:20:06 ERROR RBackendHandler: filter on 16 failed
Error in invokeJava(isStatic = FALSE, objId$id, methodName, ...) :
  org.apache.spark.sql.AnalysisException: cannot resolve 'NA' given input columns area, peri, shape, perm;

### Here there is no notion of 'NA' even though the user can see it right there

> a <- filter(df, "isnull(peri)")
> head(a)
  area peri    shape perm
1 7558   NA 0.183312  6.3
> a <- filter(df, "isnotnull(peri)")
> head(a)
  area    peri     shape perm
1 4990 2791.90 0.0903296  6.3
2 7002 3892.60 0.1486220  6.3
3 7352 3869.32 0.1170630  6.3
4 7943 3948.54 0.1224170 17.1
5 7979 4010.15 0.1670450 17.1
6 9333 4345.75 0.1896510 17.1
> a <- filter(df, "peri IS NULL")
> head(a)
  area peri    shape perm
1 7558   NA 0.183312  6.3
```

If we are to keep the automatic NULL<->NA conversion then I'd suggest we don't expose a `isnull` `isnotnull`

Comment (felixcheung) [inline R/pkg/R/generics.R]: also please note my comment earlier-  isNaN exists before 1.6 - if we remove it, we should explain this in release note.

File: R/pkg/R/generics.R [hunk 5]
```diff
@@ -796,9 +800,17 @@ setGeneric("initcap", function(x) { standardGeneric("initcap") })
 #' @export
 setGeneric("instr", function(y, x) { standardGeneric("instr") })
 
-#' @rdname isNaN
+#' @rdname is.nan
 #' @export
-setGeneric("isNaN", function(x) { standardGeneric("isNaN") })
+setGeneric("is.nan")
```
Comment (sun-rui) [inline R/pkg/R/generics.R]: it seems that "is.nan" in base package is masked? There is an implicit generic function for a primitive function. I think no need to define it here.

File: R/pkg/R/generics.R (hunk 4 above)
Comment (yanboliang) [inline R/pkg/R/generics.R]: @felixcheung I agree to not expose ```isnull``` and ```isnotnull```. Could you tell me how to update the release note to show ```isNaN``` is a breaking change?

File: R/pkg/R/generics.R (hunk 5 above)
Comment (yanboliang) [inline R/pkg/R/generics.R]: OK, I will update it.

File: R/pkg/R/generics.R (hunk 4 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: on rel note: see https://issues.apache.org/jira/browse/SPARK-11238

File: R/pkg/R/generics.R (hunk 5 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: please also add a test for base::is.nan if we are masking it. See https://github.com/yanboliang/spark/blob/spark-12044/R/pkg/inst/tests/test_sparkSQL.R#L931

**Comment by sun-rui** [comment]
LGTM

File: R/pkg/R/generics.R (hunk 4 above)
Comment (yanboliang) [inline R/pkg/R/generics.R]: open https://issues.apache.org/jira/browse/SPARK-12118 to explain in release note.

Comment (shivaram) [inline R/pkg/R/generics.R]: I don't understand the resolution here? `isNull` is still exported after this PR ? How is @felixcheung 's example going to change after this PR ?

Comment (yanboliang) [inline R/pkg/R/generics.R]: ```isNull``` and ```isNotNull``` are still exported after this PR because they have been exist at Spark 1.5. ```isNull``` and ```isNotNull``` (with upper case) are functions of ```Column```, I consider  they are Spark specific functions so did not remove them. If we want to remove them, I can do it in a follow-up PR and they are breaking changes need to be explain in release note. @shivaram

Comment (shivaram) [inline R/pkg/R/generics.R]: Ok I think I get it. Let me summarize the situation below and let know if I am getting it right.

1. We have isNaN, isNull, isNotNull for `Column` as defined in `column.R`. These mirror the scala functions.
2. We have added `isnan` and `is.nan` for `Column` in this PR. These call `isnan` in Scala. And I presume their behavior is this the same as `isNaN` ?
3. In addition to this, we have some DataFrame operators called `isNaN` ? I can't find that call in our unit test file, so I guess it doesn't exist in SparkR ? Does this exist in Scala ?
4. We convert `NA` in R to `null` in the SparkSQL side. 

I think the change looks fine to me, but I just want to understand the different things going on here

Comment (felixcheung) [inline R/pkg/R/generics.R]: I believe that's correct.We could scope this PR to only isnan on Column. And track others with JIRAs.

Comment (sun-rui) [inline R/pkg/R/generics.R]: @shivaram, to be clear, IsNaN, isnan and is.nan are functions (in org.apache.spark.sql.functions) applied to Column, they are not DataFrame operators. IsNaN function is deprecated by isnan function. We add is.nan as an alias of isnan.

**Comment by shivaram** [comment]
Thanks @sun-rui @felixcheung for the clarifications and @yanboliang for the change. LGTM. Merging this to master, branch-1.6

**Comment by yanboliang** [comment]
Thanks for your help @sun-rui @felixcheung @shivaram .
